File: rust/src/terminal.rs
use std::mem;

use libc;
use libc::termios as Termios;

use crate::error::TapestryError;
use crate::escape;
use crate::io::write;

pub fn set_up(
    allow_private: bool,
) -> Result<impl FnOnce() -> Result<(), TapestryError>, TapestryError> {
    eprintln!("running terminal set up");

    if allow_private {
        write(escape::PRIVATE_SAVE_SCREEN)?;
        write(escape::PRIVATE_SAVE_CURSOR)?;
        write(escape::PRIVATE_HIDE_CURSOR)?;
    }

    let original_termios = enable_raw_mode()?;
    reset()?;

    Ok(move || tear_down(original_termios, allow_private))
}

pub fn get_dimensions() -> Result<(u16, u16), TapestryError> {
    let mut size = libc::winsize {
        ws_row: 0,
        ws_col: 0,
        ws_xpixel: 0,
        ws_ypixel: 0,
    };

    let result = unsafe {
        libc::ioctl(libc::STDOUT_FILENO, libc::TIOCGWINSZ.into(), &mut size)
    };
    if result == -1 {
        return Err(TapestryError::new("ioctl() failed".to_string()));
    }

    Ok((size.ws_row, size.ws_col))
}

fn tear_down(
    original: Termios,
    allow_private: bool,
) -> Result<(), TapestryError> {
    eprintln!("running terminal tear down");

    if allow_private {
        write(escape::PRIVATE_SHOW_CURSOR)?;
        write(escape::PRIVATE_RESTORE_CURSOR)?;
        write(escape::PRIVATE_RESTORE_SCREEN)?;
    } else {
        reset()?;
    }

    disable_raw_mode(original)?;

    Ok(())
}

fn enable_raw_mode() -> Result<Termios, TapestryError> {
    let original: Termios = unsafe {
        let mut original = mem::zeroed();
        let result = libc::tcgetattr(libc::STDOUT_FILENO, &mut original);
        if result < 0 {
            return Err(TapestryError::new("tcgetattr() failed".to_string()));
        }

        original
    };

    let mut copy = original.clone();
    copy.c_iflag &=
        !(libc::BRKINT | libc::ICRNL | libc::INPCK | libc::ISTRIP | libc::IXON);
    copy.c_oflag &= !libc::OPOST;
    copy.c_cflag |= libc::CS8;
    copy.c_lflag &= !(libc::ECHO | libc::ICANON | libc::IEXTEN | libc::ISIG);

    copy.c_cc[libc::VMIN] = 0; // min bytes read() nees to return
    copy.c_cc[libc::VTIME] = 1; // max read() wait time in 10ths of a second

    unsafe {
        let result =
            libc::tcsetattr(libc::STDOUT_FILENO, libc::TCSADRAIN, &copy);

        if result < 0 {
            return Err(TapestryError::new("tcsetattr() failed".to_string()));
        }
    }

    eprintln!("raw mode enabled");
    Ok(original)
}

fn disable_raw_mode(original: Termios) -> Result<(), TapestryError> {
    unsafe {
        let result =
            libc::tcsetattr(libc::STDOUT_FILENO, libc::TCSADRAIN, &original);

        if result < 0 {
            return Err(TapestryError::new("tcsetattr() failed".to_string()));
        }
    }

    eprintln!("raw mode disabled");
    Ok(())
}

fn reset() -> Result<(), TapestryError> {
    write(escape::ERASE_ALL)?;
    write(escape::CURSOR_HOME)?;
    Ok(())
}

File: zig/src/terminal.zig
const std = @import("std");
const c = @cImport(
    @cInclude("termios.h"),
);
const escape = @import("escape.zig");

var original_termios: std.posix.termios = undefined;

pub fn set_up(allow_private: bool) !void {
    std.log.debug("{s}", .{"Running terminal set up"});

    if (allow_private) {
        try write(escape.private.save_screen);
        try write(escape.private.save_cursor);
        try write(escape.private.hide_cursor);
    }

    try enable_raw_mode();
    try reset();
}

pub fn tear_down(allow_private: bool) !void {
    std.log.debug("{s}", .{"Running terminal tear down"});

    if (allow_private) {
        try write(escape.private.show_cursor);
        try write(escape.private.restore_cursor);
        try write(escape.private.restore_screen);
    } else {
        try reset();
    }

    try disable_raw_mode();
}

pub fn write(bytes: []const u8) !void {
    try std.io.getStdOut().writeAll(bytes);
}

pub fn read_key(key: *u8) !usize {
    var buf: [1]u8 = undefined;
    const n_read = try std.io.getStdOut().read(&buf);

    if (n_read > 0) {
        key.* = buf[0];
    }

    return n_read;
}

fn reset() !void {
    try write(escape.erase_all);
    try write(escape.cursor_home);
}

fn enable_raw_mode() !void {
    original_termios = try std.posix.tcgetattr(std.io.getStdOut().handle);

    var copy = original_termios;
    copy.iflag.BRKINT = false;
    copy.iflag.ICRNL = false;
    copy.iflag.INPCK = false;
    copy.iflag.ISTRIP = false;
    copy.iflag.IXON = false;
    copy.oflag.OPOST = false;
    copy.cflag.CSIZE = .CS8;
    copy.lflag.ECHO = false;
    copy.lflag.ICANON = false;
    copy.lflag.IEXTEN = false;
    copy.lflag.ISIG = false;
    copy.cc[c.VMIN] = 0; // min bytes read() needs to return
    copy.cc[c.VTIME] = 1; // max wait time for read() in 10ths of a second

    try std.posix.tcsetattr(
        std.io.getStdOut().handle,
        .DRAIN,
        copy,
    );
}

fn disable_raw_mode() !void {
    try std.posix.tcsetattr(
        std.io.getStdOut().handle,
        .DRAIN,
        original_termios,
    );
}
